diff --git a/src/auto/configure b/src/auto/configure
index 18836113b..2a65aceb6 100755
--- a/src/auto/configure
+++ b/src/auto/configure
@@ -4984,10 +4984,8 @@ rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
 $as_echo "$cf_cv_ebcdic" >&6; }
 case "$cf_cv_ebcdic" in  #(vi
     yes)	$as_echo "#define EBCDIC 1" >>confdefs.h
-
-		line_break='"\\n"'
-		;;
-    *)		line_break='"\\012"';;
+          line_break='"\\n"';;
+    *) line_break='"\\n"';;
 esac
 
 
