diff --git a/src/os_unix.c b/src/os_unix.c
index 18c766d24..b46fd129e 100644
--- a/src/os_unix.c
+++ b/src/os_unix.c
@@ -1359,7 +1359,7 @@ mch_init(void)
 
     out_flush();
 
-#ifdef SIGTSTP
+#if defined(SIGTSTP) && !defined(__MVS__)
     // Check whether we were invoked with SIGTSTP set to be ignored. If it is
     // that indicates the shell (or program) that launched us does not support
     // tty job control and thus we should ignore that signal. If invoked as a
