diff --git a/src/auto/configure b/src/auto/configure
index bcb40a782..a98d2979d 100755
--- a/src/auto/configure
+++ b/src/auto/configure
@@ -4984,10 +4984,8 @@ rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
 $as_echo "$cf_cv_ebcdic" >&6; }
 case "$cf_cv_ebcdic" in  #(vi
     yes)	$as_echo "#define EBCDIC 1" >>confdefs.h
-
-		line_break='"\\n"'
-		;;
-    *)		line_break='"\\012"';;
+          line_break='"\\n"';;
+    *) line_break='"\\n"';;
 esac
 
 
@@ -12054,7 +12052,7 @@ int main()
 {char *s; s=(char *)tgoto("%p1%d", 0, 1); exit(!strcmp(s==0 ? "" : s, "1")); }
 
 _ACEOF
-if ac_fn_c_try_run "$LINENO"; then :
+if ac_fn_c_try_run "$LINENO" && test "$zOSUnix" = "no"; then :
 
       vim_cv_terminfo=no
 
