diff --git a/src/fileio.c b/src/fileio.c
index 180fe3906..705a9ea1a 100644
--- a/src/fileio.c
+++ b/src/fileio.c
@@ -485,6 +485,9 @@ readfile(
 				mch_access((char *)fname, W_OK))
 	    file_readonly = TRUE;
 	fd = mch_open((char *)fname, O_RDONLY | O_EXTRA, 0);
+#ifdef __MVS__
+  curbuf->needs_conversion = isdataset(fd) || __file_needs_conversion(fd);
+#endif
 #else
 	if (!newfile
 		|| readonlymode
