diff --git a/src/vim.h b/src/vim.h
index 2bafda424..4a899fe13 100644
--- a/src/vim.h
+++ b/src/vim.h
@@ -1837,8 +1837,14 @@ typedef void	    *vim_acl_T;		// dummy to pass an ACL to a function
 # define vim_read(fd, buf, count)   read((fd), (char *)(buf), (unsigned int)(count))
 # define vim_write(fd, buf, count)  write((fd), (char *)(buf), (unsigned int)(count))
 #else
-# define vim_read(fd, buf, count)   read((fd), (char *)(buf), (size_t) (count))
-# define vim_write(fd, buf, count)  write((fd), (char *)(buf), (size_t) (count))
+# ifdef __MVS__
+#   define vim_read(fd, buf, count)   zos_read((fd), (char *)(buf), (size_t) (count))
+#   define vim_write(fd, buf, count)  zos_write((fd), (char *)(buf), (size_t) (count))
+#   define close(fd)                  zos_close((fd))
+# else
+#   define vim_read(fd, buf, count)   read((fd), (char *)(buf), (size_t) (count))
+#   define vim_write(fd, buf, count)  write((fd), (char *)(buf), (size_t) (count))
+# endif
 #endif
 
 /*
@@ -2397,7 +2403,7 @@ typedef int (*opt_expand_cb_T)(optexpand_T *args, int *numMatches, char_u ***mat
 
 // This must come after including proto.h.
 // For VMS this is defined in macros.h.
-#if !defined(MSWIN) && !defined(VMS)
+#if !defined(MSWIN) && !defined(VMS) && !defined(__MVS__)
 # define mch_open(n, m, p)	open((n), (m), (p))
 # define mch_fopen(n, p)	fopen((n), (p))
 #endif
