diff --git a/src/term.c b/src/term.c
index 9c287417f..74d3743b0 100644
--- a/src/term.c
+++ b/src/term.c
@@ -2338,12 +2338,21 @@ term_7to8bit(char_u *p)
 {
     if (*p == ESC)
     {
+#ifdef __MVS__
+	if (p[1] == 0xAD)
+	    return CSI;
+	if (p[1] == 0xBD)
+	    return OSC;
+	if (p[1] == 0xD6) 
+	    return 0x8f;
+#else
 	if (p[1] == '[')
 	    return CSI;
 	if (p[1] == ']')
 	    return OSC;
 	if (p[1] == 'O')
 	    return 0x8f;
+#endif
     }
     return 0;
 }
