diff --git i/src/ex_cmds.c w/src/ex_cmds.c
index 5d9b722..c349e7d 100644
--- i/src/ex_cmds.c
+++ w/src/ex_cmds.c
@@ -14,6 +14,8 @@
 #include "vim.h"
 #include "version.h"
 
+#include "dio.h"
+
 #include <float.h>
 
 static int linelen(int *has_tab);
@@ -2862,6 +2864,10 @@ do_ecmd(
 	}
     }
 
+#ifdef __MVS__
+  if (is_dataset(sfname))
+    ffname = sfname;
+#endif
     /*
      * If the file was changed we may not be allowed to abandon it:
      * - if we are going to re-edit the same file
@@ -2880,7 +2886,6 @@ do_ecmd(
 	goto theend;
     }
 
-
     // End Visual mode before switching to another buffer, so the text can be
     // copied into the GUI selection buffer.
     // Careful: may trigger ModeChanged() autocommand
