diff --git a/src/structs.h b/src/structs.h
index 79f415d03..51d3a10b0 100644
--- a/src/structs.h
+++ b/src/structs.h
@@ -106,6 +106,10 @@ typedef struct {
  */
 #include "regexp.h"
 
+#ifdef __MVS__
+#include "dio.h"
+#endif
+
 /*
  * This is here because gui.h needs the pos_T and win_T, and win_T needs gui.h
  * for scrollbar_T.
@@ -3038,7 +3042,6 @@ struct file_buffer
 				// b_ffname
     char_u	*b_fname;	// current file name, points to b_ffname or
 				// b_sfname
-
 #ifdef UNIX
     int		b_dev_valid;	// TRUE when b_dev has a valid number
     dev_t	b_dev;		// device number
@@ -3510,6 +3513,13 @@ struct file_buffer
 #ifdef FEAT_DIFF
     int		b_diff_failed;	// internal diff failed for this buffer
 #endif
+#ifdef __MVS__
+    int needs_conversion;
+    int is_dataset;
+    int reclen;
+    char * dsname;
+#endif
+
 }; // file_buffer
 
 
