diff --git a/src/structs.h b/src/structs.h
index 36339c434..1bcc91974 100644
--- a/src/structs.h
+++ b/src/structs.h
@@ -3035,6 +3035,12 @@ struct file_buffer
 				// b_ffname
     char_u	*b_fname;	// current file name, points to b_ffname or
 				// b_sfname
+#ifdef __MVS__
+    int needs_conversion;
+    int is_dataset;
+    char_u *dsname;
+    //DFILE* ds_info;
+#endif
 
 #ifdef UNIX
     int		b_dev_valid;	// TRUE when b_dev has a valid number
